packages/wallet/api/src/sharingUnserializer.js

```diff
@@ -0,0 +1,81 @@
+// @ts-check
+import { Far, makeMarshal } from '@endo/marshal';
+
+/**
+ * Make a pair of unserializers that share low-privilege values.
+ *
+ * The "high" privilege unserializer can understand objects as slots, such as
+ * those from the wallet serializer. The "low" privilege unserializer can only
+ * understand strings as slots, such as those from the board serializer.
+ *
+ * This provides a way, for instance, to get identical brand references from
+ * both a public contract and a user's purses.
+ *
+ * @param {string} [lowProperty] - The key name to check on a high privilege
+ * slot object. If the string value at this key is equivalent to the value of
+ * another low privilege slot, the high privilege unserializer will return the
+ * same object for both slots.
+ */
+export const makeSharingUnserializer = lowProperty => {
+  assert(lowProperty, 'lowProperty must be specified for sharing unserializer');
+  const seen = {
+    // high: new WeakMap(),  // ISSUE: WeakMap conflicts with string keys
+    high: new Map(),
+    low: new Map(),
+  };
+
+  const slotToVal = {
+    /**
+     * @param {Record<string, any>} slot
+     * @param {string} iface
+     */
+    high: (slot, iface) => {
+      let obj;
+      let key;
+
+      if (lowProperty in slot) {
+        key = slot[lowProperty];
+        if (seen.low.has(key)) {
+          return seen.low.get(key);
+        }
+        obj = Far(iface, {});
+        seen.low.set(key, obj);
+      } else {
+        key = JSON.stringify(slot);
+        if (seen.high.has(key)) {
+          return seen.high.get(key);
+        }
+        obj = Far(iface, {});
+        seen.high.set(key, obj);
+      }
+      return obj;
+    },
+
+    /**
+     * @param {Record<string, any>} slot
```

Comment:
If the `low` serializer is meant for data coming from the AMM (and serialized by the Board's marshaller, whose emitted slots are boardID strings), then maybe this signature should be `@param {string} slot` instead of a Record, which would address the concern below.